docs/tutorials/step-8.md

```diff
@@ -0,0 +1,172 @@
+
+### A complete example
+
+Let's write an exercise dedicated to *Peano numbers*. Here is the structure of the exercise:
+
+```
+.
+├── exercises
+│   ├── index.json
+│   └── lib
+│   │   ├── check.ml
+│   │   └── check.mli
+│   ├── peano
+│   │   ├── depend.txt
+│   │   ├── descr.md
+│   │   ├── meta.json
+│   │   ├── prelude.ml
+│   │   ├── prepare.ml
+│   │   ├── solution.ml
+│   │   ├── template.ml
+│   │   ├── test.ml
+│   │   └── tests
+│   │   │   ├── samples.ml
+│   │       ├── add.ml
+│   │       └── odd_even.ml
+│   ├── an-other-exercise
+│   │   ├── depend.txt
+│   │   │ ...
+```
+The exercise **peano** follows the classical format : **prelude.ml**, **prepare.ml**, **solution.ml**, **template.ml** and **test.ml**. It comprises moreover several dependencies (**check.ml**, **samples.ml**, **add.ml** and **odd_even.ml**) which are declared as follow in **depend.txt**:
+
+```txt
+../lib/check.mli
+../lib/check.ml    # a comment
+
+tests/samples.ml
+tests/add.ml
+tests/odd_even.ml
+``` 
+
+Here is in details the source code of the exercise :
+
+- **descr.md**
+
+> * implement the function `add : peano -> peano -> peano` ; 
+> * implement the functions `odd : peano -> bool` and `even : peano -> bool`.
+
+- **prelude.ml**
+```ocaml
+type peano = Z | S of peano
+```
+
+- **solution.ml**
+```ocaml
+let rec add n = function
+| Z -> n
+| S m -> S (add n m)
+
+let rec odd = function
+| Z -> false 
+| S n -> even n
+and even = function
+| Z -> true
+| S n -> odd n
+```
+
+- **test.ml**
+```ocaml
+let () =
+  Check.safe_set_result [ Add.test ; Odd_even.test ]
+```
+
+Note that **test.ml** is very compact because it simply combines functions defined in separated files.
+
+- **../lib/check.ml**:
+```ocaml
+open Test_lib
+open Report
+
+let safe_set_result tests =
+  set_result @@
+  ast_sanity_check code_ast @@ fun () ->
+    List.mapi (fun i test -> 
+                Section ([ Text ("Question " ^ string_of_int i ^ ":") ],
+                         test ())) tests
+```
+- **../lib/check.mli**:
+```ocaml
+val safe_set_result : (unit -> Report.t) list -> unit
+```
+
+- **tests/add.ml**: 
+```ocaml
+let test () =
+  Test_lib.test_function_2_against_solution
+    [%ty : peano -> peano -> peano ] "add"
+    [ (Z, Z) ; (S(Z), S(S(Z))) ]
+```
+- **tests/odd_even.ml** :
+```ocaml
+let test () =
+  Test_lib.test_function_1_against_solution
+    [%ty : peano -> bool ] "odd"
+    [ Z ; S(Z) ; S(S(Z)) ] 
+  @
+  Test_lib.test_function_1_against_solution
+    [%ty : peano -> bool ] "even"
+    [ Z ; S(Z) ; S(S(Z)) ]
+```
+Remember that **Test_lib** internally requires a user-defined sampler `sample_peano : unit -> peano` to generate value of type `peano`. This sampler has to be present in the toplevel environment -- and not in a module -- in order to be found by the introspection primitives during grading. Therefore, we define this sampler in a file starting by the annotation `@included`.
+- **samples.ml**:
+```ocaml
+@included
```

Comment:
Using the standard syntax:
```
[@@@included]
```
would have the benefits of making this .ml file accepted by the compiler.

Reply:
I have switched to the standard syntax.

Note that the parsing of the annotation is hard-coded [in grading.ml](https://github.com/lsylvestre/learn-ocaml/blob/e99fbc9d9e020d9abe6d641c5534b40b6e548111/src/grader/grading.ml#L167) as a string comparison:
 ```ocaml
match String.trim (String.sub content 0 i) with 
| "[@@@included]" -> ...
```

Is this the right way to do it?

Reply:
That's probably fine.

A cleaner way would be to inspect the module syntax tree but I am not sure if that is easy for you to get this tree at this point of the process.